TLSv1.3: Add Early Data support

mattcaswell opened the pull request:
<!--
Thank you for your pull request. Please review below requirements.

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated
- [x] tests are added or updated

##### Description of change
<!-- Provide a description of the changes.

If it fixes a github issue, add Fixes #XXXX.
-->

This adds support for TLSv1.3 early data.

There are a number of new functions added. The key ones are SSL_read_early() and SSL_write_early(). These take the same parameters as SSL_read_ex() and SSL_write_ex(), but must be called first on a connection, i.e. before any calls to SSL_connect()/SSL_do_handshake()/SSL_read()/SSL_write(). I have provided a fuller description in the docs in the PR.

I've also added "-early_data" options to s_client and s_server to test it out. The s_client version takes a filename as an argument, and sends the contents of that file as early data to the server.

Marked as WIP because this still needs tests.

richsalz commented:
Is there an API to tell if there's early data and how much?
What happens if a client sends early data and the server just does a read; is it dropped (I hope so)

mattcaswell commented:
@Lukasa, @Lekensteyn - not really tested that aspect but this PR should add the early secret logging.

mattcaswell commented:
> Is there an API to tell if there's early data and how much?

Well you don't know if you're going to get early data until we've received the ClientHello. If you wait until after the connection has completed and then ask "have I got early data to read?" its too late. We would have had to cache all that early data somewhere - and then hang around waiting for the handshake to complete, before continuing (which is kind of against the point). Instead you attempt to start reading early data immediately and then get told if there was any or not. The implementation drives the handshake through to the point where the server sends back its Finished (but before the client responds with its own Finished), and then streams the early data to the server app.

> What happens if a client sends early data and the server just does a read; is it dropped (I hope so)

Yes.

Lekensteyn commented on ssl/record/ssl3_record.c:
```diff
@@ -101,16 +101,16 @@ static int ssl3_record_app_data_waiting(SSL *s)
-    if (s->session->ext.max_early_data < s->max_early_data)
-        max_early_data = s->max_early_data;
+    if (s->hit && s->session->ext.max_early_data < s->max_early_data)
+        max_early_data = s->session->ext.max_early_data;
```
Ah ok, previous issue (use of s->max_early_data instead of s->session->...) is fixed here.

Lekensteyn commented:
(After draft -18, ticket_early_data_info(46) is replaced by a new max_early_data_size field in the early_data(42) extension, but this should not be present for CH/EE, only NST.)

mattcaswell commented on ssl/statem/extensions_clnt.c:
```diff
@@ -108,6 +108,26 @@ static int use_ecc(SSL *s)
+
+    if (!WPACKET_put_bytes_u16(pkt, TLSEXT_TYPE_early_data)
+            || !WPACKET_start_sub_packet_u16(pkt)
+            || !WPACKET_close(pkt)) {
```
I'm not sure I understand your point. This is just constructing an empty extension as per draft 18.

mattcaswell commented on ssl/statem/extensions_srvr.c:
```diff
@@ -832,6 +843,22 @@ int tls_construct_stoc_early_data_info(SSL *s, WPACKET *pkt,
+
+    if (!WPACKET_put_bytes_u16(pkt, TLSEXT_TYPE_early_data)
+            || !WPACKET_start_sub_packet_u16(pkt)
+            || !WPACKET_close(pkt)) {
```
As above. This is just constructing an empty extension

Lekensteyn replied:
Ah, I was not familar with this API and thought that this was constructing the contents of an extension. Nevermind then, then it should be OK

mattcaswell commented:
> (After draft -18, ticket_early_data_info(46) is replaced by a new max_early_data_size field in the early_data(42) extension, but this should not be present for CH/EE, only NST.)

Right. All of this is based off draft 18, for the sake of having something stable to work from.

levitte commented on ssl/ssl_lib.c:
```diff
@@ -1544,6 +1545,14 @@ int ssl_read_internal(SSL *s, void *buf, size_t num, size_t *readbytes)
+            && s->early_data_state != SSL_EARLY_DATA_FINISHED_WRITING
+            && s->early_data_state != SSL_EARLY_DATA_FINISHED_READING
+            && s->early_data_state != SSL_EARLY_DATA_READING) {
+        SSLerr(SSL_F_SSL_WRITE_INTERNAL, ERR_R_SHOULD_NOT_HAVE_BEEN_CALLED);
```
s/SSL_F_SSL_WRITE_INTERNAL/SSL_F_SSL_READ_INTERNAL/

levitte commented on ssl/ssl_lib.c:
```diff
@@ -3077,6 +3237,16 @@ int SSL_do_handshake(SSL *s)
+            && s->early_data_state != SSL_EARLY_DATA_FINISHED_READING
+            && s->early_data_state != SSL_EARLY_DATA_ACCEPTING
+            && s->early_data_state != SSL_EARLY_DATA_CONNECTING) {
+        SSLerr(SSL_F_SSL_WRITE_INTERNAL, ERR_R_SHOULD_NOT_HAVE_BEEN_CALLED);
```
s/SSL_F_SSL_WRITE_INTERNAL/SSL_F_SSS_DO_HANDSHAKE/

richsalz commented:
could @kaduk 's early callback be enhanced to provide an "early data is available" indicator?

mattcaswell commented:
> could @kaduk 's early callback be enhanced to provide an "early data is available" indicator?

Well, you could use it to determine whether the early_data extension is present or not. But it is called too early to determine whether the early_data will be accepted (we have to have completed version negotiation, decided whether we are resuming the session, and a number of other things first).

Although, I don't really understand what the problem is you are trying to solve. You can answer the above question with the APIs as they are in this PR. They enable you to write code that does this:

1. Run the handshake until just before when we might expect to read early data
2. Find out if you have early data or not (and if so read it)
3. Continue the handshake until the end

richsalz commented:
I think some advice in the API docs about early data would be helpful.

mattcaswell commented:
Pushed some significant updates to this.

I have revised the API a bit. I removed the `SSL_write_early_finish()` function (I realised it wasn't actually needed after some feedback from @davidben). I have also enabled the ability for the client to interleave calls to `SSL_write_early()` with `SSL_read_ex()`/`SSL_read()` (once the initial call to `SSL_write_early()` has completed successfully. Similarly, on the server side, you can interleave calls to `SSL_read_early()` with calls to `SSL_write_ex()`/`SSL_write()` as required once the initial `SSL_read_early()` call has completed successfully. This enables the server to start streaming data back to an (unauthenticated) client immediately.

I have added a set of tests for all the various scenarios. I've also updated the documentation to reflect the latest API changes.

I've not yet attended to some of the review comments and I note that there is a merge conflict for me to look at. Nonetheless I think this is close enough to take this out of WIP.

tmshort commented on ssl/statem/extensions_srvr.c:
```diff
@@ -753,6 +766,32 @@ int tls_parse_ctos_psk(SSL *s, PACKET *pkt, unsigned int context, X509 *x,
     sess->ext.tick_identity = id;
+
+    now = (uint32_t)time(NULL);
+    agesec = now - (uint32_t)sess->time;
```
Will this code be used in ~22 years? Should 64-bit values be used instead?
(Of course I'm not a fan of multiplication and division in packet processing.)

mattcaswell replied:
I don't think it makes a difference - even if it wraps I think agesec still works out right? Or am I confused.

tmshort commented on doc/man3/SSL_read_early.pod:
```diff
@@ -0,0 +1,209 @@
+SSL_CTX_get_max_early_data,
+SSL_SESSION_get_max_early_data,
+SSL_write_early,
+SSL_read_early,
```
I would suggest that these two be "SSL_write_early_data" and "SSL_read_early_data"; mostly to be consistent with the rest of the API.

richsalz replied:
I agree with @tmshort on the function names.

mattcaswell replied:
Names changed.

tmshort commented on doc/man3/SSL_read_early.pod:
```diff
@@ -0,0 +1,209 @@
+
+[[TODO(TLS1.3): The server uses SSL_write_ex()/SSL_write() to send data to an
+unauthenticated client. Should we create a separate function for this to avoid
+accidents??]]
```
I would say yes, as you want to avoid potential accidental exposure.

mattcaswell replied:
Just to be clear this is different from sending early data - only the client does that. This is about the server sending to the client while the client is still sending early data. So the server has sent all its handshake messages, but has yet to receive the authentication handshake messages from the client, i.e. the client is unauthenticated. Perhaps a suitable function name might be: SSL_write_unauth()?

richsalz replied:
If we think of the concept as "data before the handshake is complete" then ```SSL_write_early_data``` is good.  And unauth is bad for the reasons Todd said.

tmshort replied:
+1

tmshort commented on doc/man3/SSL_read_early.pod:
```diff
@@ -0,0 +1,209 @@
+bytes written must not exceed the value returned from
+SSL_SESSION_get_max_early_data(). Once the initial SSL_write_early() call has
+completed successfully the client may interleave calls to L<SSL_read_ex(3)> and
+L<SSL_read(3)> with calls to SSL_write_early() as required.
```
Should these reference SSL_read_early()? Same reason for using SSL_write_early() for the server; you want to know the security parameters of the data you're reading.

mattcaswell replied:
Well, here the client is communicating with a fully authenticated server - its not going to get any more messages from the server. Actually even after the client finishes its side of the handshake it is not going to know whether the records it is reading were sent by the server before or after it authenticated the client. That's just the way the protocol works. So I think SSL_read/SSL_read_ex is correct here.

mattcaswell commented on apps/s_server.c:
```diff
@@ -1500,7 +1506,16 @@ int s_server_main(int argc, char *argv[])
+                goto end;
+            }
+            break;
+        case OPT_EARLY_DATA:
```
No, they do different and independent things. Each makes sense by itself. If you supply "-early_data" by itself then it just uses the default value for max early data. If you supply "-max_early_data" by itself then it will reject early data up to the specified number of max early data bytes (plus an allowance for the crypto overhead). If it receives more data than that then it aborts the handshake.

richsalz replied:
reject up to?  Is that a mistake in your comment?
So -max_early_data implies early_data, right?

mattcaswell replied:
No - max_early_data has an effect whether or not we accept or reject the early data. Either way it tells us how much data the client is allowed to send. If we reject then we're going to do trial decryption until we find a record that we can read and skip over the rest. If we've done trial decryption for max_early_data bytes and still haven't found a record we can read then we abort the handshake. If you don't also supply the "-early_data" flag then we are always going to reject the early data.

richsalz replied:
that's tricky; make sure it shows up in the docs :)

mattcaswell commented on apps/s_server.c:
```diff
@@ -2191,6 +2211,45 @@ static int sv_body(int s, int stype, unsigned char *context)
+        size_t readbytes;
+
+        while (edret != SSL_READ_EARLY_FINISH) {
+            for (;;) {
```
Hmm. I thought our standard was for no spaces here. I just did a regex search and found 151 instances without spaces and 4 with - so I think the no spaces have it.

richsalz replied:
yes, you're right then.

mattcaswell commented on include/openssl/ssl.h:
```diff
@@ -894,7 +899,9 @@ typedef enum {
-    TLS_ST_CR_KEY_UPDATE
+    TLS_ST_CR_KEY_UPDATE,
+    TLS_ST_EARLY_DATA,
+    TLS_ST_PENDING_EARLY_DATA_END
```
Its values are used in the SSL_get_state() public API function.

richsalz replied:
then you have to make them int's as we've discussed previously.

mattcaswell replied:
Can't. Its already in the public API in 1.1.0 :-(

richsalz replied:
it's a bugfix and we'll backport it to 1.1.0 if needed.  this API is broken.  it's just like opacity.  it must be fixed.

mattcaswell replied:
It's only actually a problem if the values go >MAX_INT right? Which seems quite unlikely. Changing it in 1.1.0 now could break stuff - so it doesn't seem justified to me. Probably a discussion for the team rather than here though.

richsalz replied:
okay, will do so.

mattcaswell commented on ssl/t1_trce.c:
```diff
@@ -794,6 +798,15 @@ static int ssl_print_extension(BIO *bio, int indent, int server,
+        max_early_data = (ext[0] << 24) | (ext[1] << 16) | (ext[2] << 8)
+                         | ext[3];
+        BIO_indent(bio, indent + 2, 80);
+        BIO_printf(bio, "max_early_data=%d\n", max_early_data);
```
Fixed

mattcaswell commented on apps/s_server.c:
```diff
@@ -1500,7 +1506,16 @@ int s_server_main(int argc, char *argv[])
-
+        case OPT_MAX_EARLY:
+            max_early_data = atoi(opt_arg());
+            if (max_early_data < 0) {
```
It doesn't seem to? If I remove this, it lets me set a -ve value

mattcaswell replied:
I actually used 'p' originally but changed it to 'n' because I wanted to specifically allow 0 as a value. A 'p' must be > 0.

richsalz replied:
oh, okay.  then the if test stays.  do the doc update and I'll approve

mattcaswell commented on apps/s_server.c:
```diff
@@ -2076,6 +2094,8 @@ static void print_stats(BIO *bio, SSL_CTX *ssl_ctx)
                SSL_CTX_sess_get_cache_size(ssl_ctx));
 }
 
+static void print_connection_info(SSL *con);
```
Done

mattcaswell commented on doc/man3/SSL_read_early.pod:
```diff
@@ -0,0 +1,209 @@
+to support it.
+
+Early data has weaker security properties than other data sent over an SSL/TLS
+connection. In particular the data is not forward secret and the server has no
```
Done

mattcaswell commented on doc/man3/SSL_read_early.pod:
```diff
@@ -0,0 +1,209 @@
+connection. In particular the data is not forward secret and the server has no
+guarantees that the same early data was not replayed across multiple
+connections. For this reason extreme care should be exercised when using early
+data.
```
I considered it, but chose not to. I'd prefer to avoid being too specific as it will be quite dependant on the protocol I imagine.

richsalz replied:
The TLS spec itself mentioned idempotency.  And the HTTP GET is an example that almost everyone will be familiar with.  Please reconsider.

Lukasa replied:
I'd be nervous about being too specific in terms of content here. For example, a HTTP GET is *not* necessarily safe here if it contains any form of data that should be considered sensitive (e.g. cookies, authorization headers, auth tokens). A simple HTTP `GET /` to a non-authed site is likely to be fine, but for anything else the lack of forward-secrecy may be a concern.

mattcaswell replied:
I can't find it mentioned in draft -18 (search for idem returned nothing). Can you point me to where it says that. I could quote the draft.

mattcaswell commented on doc/man3/SSL_read_early.pod:
```diff
@@ -0,0 +1,209 @@
+unauthenticated client. Should we create a separate function for this to avoid
+accidents??]]
+
+On the client side the function SSL_SESSION_get_max_early_data() can be used to
```
Done

mattcaswell commented on doc/man3/SSL_read_early.pod:
```diff
@@ -0,0 +1,209 @@
+accidents??]]
+
+On the client side the function SSL_SESSION_get_max_early_data() can be used to
+determine whether a session established with a server can be used to send early
```
Done

mattcaswell commented on doc/man3/SSL_read_early.pod:
```diff
@@ -0,0 +1,209 @@
+
+A client uses the function SSL_write_early() to send early data. This function
+works in the same way as the L<SSL_write_ex(3)> function, but with a few
+differences. Refer to the L<SSL_write_ex(3)> documentation for
```
Done

mattcaswell commented on doc/man3/SSL_read_early.pod:
```diff
@@ -0,0 +1,209 @@
+works in the same way as the L<SSL_write_ex(3)> function, but with a few
+differences. Refer to the L<SSL_write_ex(3)> documentation for
+information on how to write bytes to the underlying connection, and how to
+handle any errors that may arise. This page will detail the differences between
```
Done

mattcaswell commented on include/openssl/ssl.h:
```diff
@@ -1020,6 +1027,7 @@ DECLARE_PEM_rw(SSL_SESSION, SSL_SESSION)
 # define SSL_AD_INTERNAL_ERROR           TLS1_AD_INTERNAL_ERROR
 # define SSL_AD_USER_CANCELLED           TLS1_AD_USER_CANCELLED
 # define SSL_AD_NO_RENEGOTIATION         TLS1_AD_NO_RENEGOTIATION
+# define SSL_AD_END_OF_EARLY_DATA        TLS13_AD_END_OF_EARLY_DATA
```
All of the alerts are done this way. I have no idea why. They have a protocol specific version "SSL3_AD_*", "TLS1_AD_*", "TLS13_AD_*" and then a generic name "SSL_AD_*". They are in a public header file so we can't change that. It would be odd to treat this one alert differently.

mattcaswell commented:
I rebased this, and made the updates as requested above (or otherwise responded to the comment).

mattcaswell commented:
I added a new commit to say "For specific details, consult the TLS 1.3 specification." as discussed above. I also tweaked the s_server docs to mention that early_max_data must be an integer which is 0 or greater.

richsalz commented:
nice job!

mattcaswell commented:
Yay!! Thanks. Pushed. :-)
